Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2023 Mohanad ahmed. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mohanad Ahmed
+-/
+
+import Mathlib.Data.Real.Basic
+import Mathlib.Data.Complex.Basic
+import Mathlib.Data.Matrix.Basic
+import Mathlib.Data.Rat.Basic
+import Mathlib.Data.IsROrC.Basic
+import Mathlib.LinearAlgebra.Matrix.ToLin
+
+/-! # Types with Star Inner Product such that (star v) ⬝ᵥ v = 0 ↔ v = 0
+
+This file defines StarDotProductSpace in which the star operation on a vector makes the star
+dotProduct (in 1st arg) is zero if and only if the vector is zero. This makes star operation for
+matrices of the same underlying type kernel preserving for linear maps.
+-/
+
+open BigOperators Matrix
+
+class StarDotProductSpace (n K) [Fintype n][Ring K][StarRing K] : Prop where
+  dotProduct_star_self_eq_zero (v : n → K) : Matrix.dotProduct (star v) v = 0 ↔ v = 0
```

Comment:
```suggestion
  eq_zero_of_dotProduct_star_self_eq_zero (v : n → K) : Matrix.dotProduct (star v) v = 0 → v = 0
```
As the other direction is always true, so you don't need to prove it every time.

Reply:
I used this one at first but I am reverting it.

There is no saving coming from not proving it every time (in fact I needed to add `imp_self` every where.
In addition now the lemma cannot work as a rewrite.

Reply:
The point is that you write this simpler version in the `class`, and then prove the obvious `iff` as a lemma and use _that_ as the rewrite.